### DISCUSSION TITLE
Mmap of GPIO registers failed

### DISCUSSION THREAD

**Discussion by baterflyrity** [gpiozero/gpiozero#996] [Q&A]
Operating system: debian 11 x64
Python version: 3.9
Pi model: 3 B+
GPIO Zero version: 1.6.2
Pin factory used: RPiGPIO (0.7.1), pigpio (1.78), native.

When binding (running ptyhon as root) button **mmap error** occurs independently of pin factory.

```python
>>> from gpiozero import Button
>>> button = Button(2)
Traceback (most recent call last):
  File "/usr/local/lib/python3.9/dist-packages/gpiozero/pins/pi.py", line 101, in pin
    pin = self.pins[n]
KeyError: 24

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/local/lib/python3.9/dist-packages/gpiozero/devices.py", line 108, in __call__
    self = super(GPIOMeta, cls).__call__(*args, **kwargs)
  File "/usr/local/lib/python3.9/dist-packages/gpiozero/input_devices.py", line 414, in __init__
    super(Button, self).__init__(
  File "/usr/local/lib/python3.9/dist-packages/gpiozero/mixins.py", line 433, in __init__
    super(HoldMixin, self).__init__(*args, **kwargs)
  File "/usr/local/lib/python3.9/dist-packages/gpiozero/input_devices.py", line 165, in __init__
    super(DigitalInputDevice, self).__init__(
  File "/usr/local/lib/python3.9/dist-packages/gpiozero/mixins.py", line 218, in __init__
    super(EventsMixin, self).__init__(*args, **kwargs)
  File "/usr/local/lib/python3.9/dist-packages/gpiozero/input_devices.py", line 83, in __init__
    super(InputDevice, self).__init__(pin, pin_factory=pin_factory)
  File "/usr/local/lib/python3.9/dist-packages/gpiozero/devices.py", line 549, in __init__
    pin = self.pin_factory.pin(pin)
  File "/usr/local/lib/python3.9/dist-packages/gpiozero/pins/pi.py", line 103, in pin
    pin = self.pin_class(self, n)
  File "/usr/local/lib/python3.9/dist-packages/gpiozero/pins/rpigpio.py", line 111, in __init__
    GPIO.setup(self.number, GPIO.IN, self.GPIO_PULL_UPS[self._pull])
RuntimeError: Mmap of GPIO registers failed
```
Also my permissions:
```bash
root@operator:~# ls -ld /sys/class/gpio/export
-rwxr-xr-x 1 root dialout 4096 Sep 24 17:44 /sys/class/gpio/export
root@operator:~# echo 2 > /sys/class/gpio/export
-bash: echo: write error: Invalid argument
```

**Comment by lurch** [comment]
> Operating system: debian 11 x64
> Pi model: 3 B+

Typo? The Pi isn't able to run x64 OSes :wink: 

> RuntimeError: Mmap of GPIO registers failed

Does (the permissions stuff in) #837 help?

**Comment by baterflyrity** [comment]
> Typo? The Pi isn't able to run x64 OSes 😉

Sry, ofc its [x32](https://wiki.debian.org/ArmHardFloatPort).

**Comment by baterflyrity** [comment]
> 
> 
> > Typo? The Pi isn't able to run x64 OSes 😉
> 
> Sry, ofc its [x32](https://wiki.debian.org/ArmHardFloatPort).

But i get:
```bash
root@debian:~# uname -a
Linux debian 5.10.0-8-arm64 #1 SMP Debian 5.10.46-4 (2021-08-03) aarch64 GNU/Linux
```

**Comment by lurch** [comment]
Going a bit off-topic, but the 'x' in x32 and x64 generally stands for x86 (i.e. the architecture used by Intel / AMD CPUs). The Pi uses ARM CPUs, which are generally referred to as armhf for 32-bit OSes, or aarch64 for 64-bit OSes. :computer: :slightly_smiling_face: (sorry for being pedantic)

**Comment by baterflyrity** [comment]
@lurch thanks for link but now the problem seems to be in no gpiomem device:

```bash
root@debian:~# ls /dev/ | grep gpio
gpiochip0
gpiochip1
```

How can i add gpiomem device?

**Reply by flsabourin** [reply]
add `iomem=relaxed ` in your cmdline.txt

**Comment by bennuttall** [comment]
You'd be better off using either Raspberry Pi OS or Ubuntu, as these both provide working GPIO interfaces. See https://waldorf.waveform.org.uk/2021/the-pins-they-are-a-changin.html for more info on how the different pin interfaces work.

**Reply by baterflyrity** [reply]
Thanks, got your hint. 

Am i right you consider *gpiozero* does support **only** Raspberry Pi OS and Ubuntu?

As stated in your link *gpiozero* has to work properly but it does not. So should i migrate to *python3-lgpio*?

**Reply by bennuttall** [reply]
gpiozero will work in any OS with a functional pin library. If you can get RPi.GPIO or pigpio or lgpio working, then gpiozero will *just work*. We *support* any OS in that gpiozero is made to run anywhere, but we don't *provide support* beyond the basics. So if you can get a pin library working on an OS, you should be able to use gpiozero easily.

You can try lgpio if you want. It's not available in Debian yet, as far as I'm aware, but you should be able to pip install it. And again, you'll need to do the prep to get it working. This might mean setting up groups with access to the GPIOs. Like I said, if you want it to work out of the box, use RPiOS or Ubuntu. If you're willing to figure out how to get it working on Debian, feel free, but we can't provide support beyond what's in the docs and what's been said here.

**Reply by baterflyrity** [reply]
Thanks. As i understand gpiozero does not support backend installation and needs it to be installed firstly manually?

**Reply by bennuttall** [reply]
I'm not sure what you mean, but yes you need to install a pin library manually (apt install or pip install).

**Reply by lurch** [reply]
In an "ideal world" I guess Debian 11 would list `lgpio` as a requirement (or recommendation?) of `gpiozero`, and in that case installing gpiozero with `apt` would "just work" :wink:

**Reply by perlun** [reply]
Late to the party, but it's definitely doable to get GPIO support working in Python on Debian, by following these instructions: https://wiki.debian.org/RaspberryPi4#GPIO_support (I just tested them myself tonight :tada:)

**Reply by waveform80** [reply]
I did have a shot at [getting lgpio into Debian](https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=990280) but after a few months of back and forth it must've slid off the bottom of my mental "let's work on this" list. I also prepped an updated [gpiozero for Debian](https://salsa.debian.org/python-team/packages/gpiozero) but obviously that's not going to go anywhere without the former. I also needed colorzero to land in Debian (another new dep of gpiozero) and that ... oh, [that made it in](https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=924643); I missed that!

Anyway, if anyone wants to pick it up, feel free to fork it on salsa; I'm afraid my plate is full (and my enthusiasm for packaging debates is pretty much zero at the moment)

**Reply by lurch** [reply]
> my enthusiasm for packaging debates is pretty much zero at the moment

You mean there's not going to be a FlatPak version of gpiozero?? :sob:  :wink: :laughing:
